### Lecture 170 - Going Multidimensional 

In [111]:
import pandas as pd

In [112]:
tech = pd.read_csv("D:\\Data for Preprocessing\\tech_giants.csv") 

In [113]:
tech

date  month  year  day   name     open    close     high      low  \
0     02-01-2014      1  2014    2     FB    54.86    54.71    55.22    54.19   
1     02-01-2014      1  2014    2   AAPL    79.38    79.02    79.58    78.86   
2     02-01-2014      1  2014    2  GOOGL   557.73   556.56   558.88   554.13   
3     02-01-2014      1  2014    2   MSFT    37.35    37.16    37.40    37.10   
4     02-01-2014      1  2014    2   AMZN   398.80   397.97   399.36   394.02   
...          ...    ...   ...  ...    ...      ...      ...      ...      ...   
7100  23-08-2019      8  2019   23   MSFT   137.19   133.39   138.35   132.80   
7101  23-08-2019      8  2019   23   AAPL   209.43   202.64   212.05   201.00   
7102  23-08-2019      8  2019   23  GOOGL  1185.17  1153.58  1195.67  1150.00   
7103  23-08-2019      8  2019   23   AMZN  1793.03  1749.62  1804.90  1745.23   
7104  23-08-2019      8  2019   23     FB   180.84   177.75   183.13   176.66   

        volume volume_type  
0     43257622      medium  
1      8398851         low  
2      1822719      medium  
3     30643745      medium  
4      2140246      medium  
...        ...         ...  
7100  38515386      medium  
7101  46882843      medium  
7102   1813141      medium  
7103   5277898      medium  
7104  17331221      medium  

[7105 rows x 11 columns]

In [114]:
tech.set_index('date')

month  year  day   name     open    close     high      low  \
date                                                                      
02-01-2014      1  2014    2     FB    54.86    54.71    55.22    54.19   
02-01-2014      1  2014    2   AAPL    79.38    79.02    79.58    78.86   
02-01-2014      1  2014    2  GOOGL   557.73   556.56   558.88   554.13   
02-01-2014      1  2014    2   MSFT    37.35    37.16    37.40    37.10   
02-01-2014      1  2014    2   AMZN   398.80   397.97   399.36   394.02   
...           ...   ...  ...    ...      ...      ...      ...      ...   
23-08-2019      8  2019   23   MSFT   137.19   133.39   138.35   132.80   
23-08-2019      8  2019   23   AAPL   209.43   202.64   212.05   201.00   
23-08-2019      8  2019   23  GOOGL  1185.17  1153.58  1195.67  1150.00   
23-08-2019      8  2019   23   AMZN  1793.03  1749.62  1804.90  1745.23   
23-08-2019      8  2019   23     FB   180.84   177.75   183.13   176.66   

              volume volume_type  
date                              
02-01-2014  43257622      medium  
02-01-2014   8398851         low  
02-01-2014   1822719      medium  
02-01-2014  30643745      medium  
02-01-2014   2140246      medium  
...              ...         ...  
23-08-2019  38515386      medium  
23-08-2019  46882843      medium  
23-08-2019   1813141      medium  
23-08-2019   5277898      medium  
23-08-2019  17331221      medium  

[7105 rows x 10 columns]

### Lecture - 172 - Creating a Multiindex

In [115]:
tech.set_index(['date', 'name'], inplace=True)

In [116]:
tech.head(10)

month  year  day    open   close    high     low    volume  \
date       name                                                                
02-01-2014 FB         1  2014    2   54.86   54.71   55.22   54.19  43257622   
           AAPL       1  2014    2   79.38   79.02   79.58   78.86   8398851   
           GOOGL      1  2014    2  557.73  556.56  558.88  554.13   1822719   
           MSFT       1  2014    2   37.35   37.16   37.40   37.10  30643745   
           AMZN       1  2014    2  398.80  397.97  399.36  394.02   2140246   
03-01-2014 FB         1  2014    3   55.00   54.56   55.65   54.53  38287706   
           GOOGL      1  2014    3  557.50  552.50  558.47  552.47   1669229   
           MSFT       1  2014    3   37.20   36.91   37.22   36.60  31134795   
           AAPL       1  2014    3   79.00   77.28   79.10   77.20  14043410   
           AMZN       1  2014    3  398.29  396.44  402.71  396.22   2213512   

                 volume_type  
date       name               
02-01-2014 FB         medium  
           AAPL          low  
           GOOGL      medium  
           MSFT       medium  
           AMZN       medium  
03-01-2014 FB         medium  
           GOOGL      medium  
           MSFT       medium  
           AAPL          low  
           AMZN       medium

In [117]:
type(tech.index)

pandas.core.indexes.multi.MultiIndex

### Lecture - 173 - Multiindex from Read_CSV

In [118]:
Demo_Data = pd.read_csv("D:\\Data for Preprocessing\\tech_giants.csv", index_col=['date', 'name']) 

In [119]:
Demo_Data

month  year  day     open    close     high      low  \
date       name                                                          
02-01-2014 FB         1  2014    2    54.86    54.71    55.22    54.19   
           AAPL       1  2014    2    79.38    79.02    79.58    78.86   
           GOOGL      1  2014    2   557.73   556.56   558.88   554.13   
           MSFT       1  2014    2    37.35    37.16    37.40    37.10   
           AMZN       1  2014    2   398.80   397.97   399.36   394.02   
...                 ...   ...  ...      ...      ...      ...      ...   
23-08-2019 MSFT       8  2019   23   137.19   133.39   138.35   132.80   
           AAPL       8  2019   23   209.43   202.64   212.05   201.00   
           GOOGL      8  2019   23  1185.17  1153.58  1195.67  1150.00   
           AMZN       8  2019   23  1793.03  1749.62  1804.90  1745.23   
           FB         8  2019   23   180.84   177.75   183.13   176.66   

                    volume volume_type  
date       name                         
02-01-2014 FB     43257622      medium  
           AAPL    8398851         low  
           GOOGL   1822719      medium  
           MSFT   30643745      medium  
           AMZN    2140246      medium  
...                    ...         ...  
23-08-2019 MSFT   38515386      medium  
           AAPL   46882843      medium  
           GOOGL   1813141      medium  
           AMZN    5277898      medium  
           FB     17331221      medium  

[7105 rows x 9 columns]

### Lecture - 174 - Indexing Hierarchical DataFrames

In [120]:
tech.head(10)

month  year  day    open   close    high     low    volume  \
date       name                                                                
02-01-2014 FB         1  2014    2   54.86   54.71   55.22   54.19  43257622   
           AAPL       1  2014    2   79.38   79.02   79.58   78.86   8398851   
           GOOGL      1  2014    2  557.73  556.56  558.88  554.13   1822719   
           MSFT       1  2014    2   37.35   37.16   37.40   37.10  30643745   
           AMZN       1  2014    2  398.80  397.97  399.36  394.02   2140246   
03-01-2014 FB         1  2014    3   55.00   54.56   55.65   54.53  38287706   
           GOOGL      1  2014    3  557.50  552.50  558.47  552.47   1669229   
           MSFT       1  2014    3   37.20   36.91   37.22   36.60  31134795   
           AAPL       1  2014    3   79.00   77.28   79.10   77.20  14043410   
           AMZN       1  2014    3  398.29  396.44  402.71  396.22   2213512   

                 volume_type  
date       name               
02-01-2014 FB         medium  
           AAPL          low  
           GOOGL      medium  
           MSFT       medium  
           AMZN       medium  
03-01-2014 FB         medium  
           GOOGL      medium  
           MSFT       medium  
           AAPL          low  
           AMZN       medium

In [121]:
# Usual Method 
tech.loc['02-01-2014', 'GOOGL'].close

556.56

In [122]:
# Passing the Tuple
tech.loc[('02-01-2014', 'GOOGL'), 'open' : 'close']

open     557.73
close    556.56
Name: (02-01-2014, GOOGL), dtype: float64

In [123]:
tech.iloc[2, 3 : 5]

open     557.73
close    556.56
Name: (02-01-2014, GOOGL), dtype: float64

### Lecture - 175 - Indexing Ranges and Slices

In [124]:
tech

month  year  day     open    close     high      low  \
date       name                                                          
02-01-2014 FB         1  2014    2    54.86    54.71    55.22    54.19   
           AAPL       1  2014    2    79.38    79.02    79.58    78.86   
           GOOGL      1  2014    2   557.73   556.56   558.88   554.13   
           MSFT       1  2014    2    37.35    37.16    37.40    37.10   
           AMZN       1  2014    2   398.80   397.97   399.36   394.02   
...                 ...   ...  ...      ...      ...      ...      ...   
23-08-2019 MSFT       8  2019   23   137.19   133.39   138.35   132.80   
           AAPL       8  2019   23   209.43   202.64   212.05   201.00   
           GOOGL      8  2019   23  1185.17  1153.58  1195.67  1150.00   
           AMZN       8  2019   23  1793.03  1749.62  1804.90  1745.23   
           FB         8  2019   23   180.84   177.75   183.13   176.66   

                    volume volume_type  
date       name                         
02-01-2014 FB     43257622      medium  
           AAPL    8398851         low  
           GOOGL   1822719      medium  
           MSFT   30643745      medium  
           AMZN    2140246      medium  
...                    ...         ...  
23-08-2019 MSFT   38515386      medium  
           AAPL   46882843      medium  
           GOOGL   1813141      medium  
           AMZN    5277898      medium  
           FB     17331221      medium  

[7105 rows x 9 columns]

In [125]:
tech.loc[(['02-01-2014', '03-01-2014'], ['GOOGL']), : ]

month  year  day    open   close    high     low   volume  \
date       name                                                               
02-01-2014 GOOGL      1  2014    2  557.73  556.56  558.88  554.13  1822719   
03-01-2014 GOOGL      1  2014    3  557.50  552.50  558.47  552.47  1669229   

                 volume_type  
date       name               
02-01-2014 GOOGL      medium  
03-01-2014 GOOGL      medium

In [126]:
tech.loc[(['02-01-2014', '03-01-2014'], ['GOOGL', 'MSFT']), : ]

month  year  day    open   close    high     low    volume  \
date       name                                                                
02-01-2014 GOOGL      1  2014    2  557.73  556.56  558.88  554.13   1822719   
           MSFT       1  2014    2   37.35   37.16   37.40   37.10  30643745   
03-01-2014 GOOGL      1  2014    3  557.50  552.50  558.47  552.47   1669229   
           MSFT       1  2014    3   37.20   36.91   37.22   36.60  31134795   

                 volume_type  
date       name               
02-01-2014 GOOGL      medium  
           MSFT       medium  
03-01-2014 GOOGL      medium  
           MSFT       medium

In [127]:
tech.loc[(['02-01-2014', '03-01-2014'], ['GOOGL', 'MSFT']), ['high', 'low'] ]

high     low
date       name                 
02-01-2014 GOOGL  558.88  554.13
           MSFT    37.40   37.10
03-01-2014 GOOGL  558.47  552.47
           MSFT    37.22   36.60

In [128]:
#  This will give an error
tech.loc[(['02-01-2014': '07-01-2014'], ['GOOGL', 'MSFT']), ['high', 'low'] ]

SyntaxError: invalid syntax (Temp/ipykernel_18176/3635494701.py, line 2)

In [ ]:
tech.loc[(['02-01-2014', '03-01-2014']), ['high', 'low'] ]

high     low
date       name                 
02-01-2014 FB      55.22   54.19
           AAPL    79.58   78.86
           GOOGL  558.88  554.13
           MSFT    37.40   37.10
           AMZN   399.36  394.02
03-01-2014 FB      55.65   54.53
           GOOGL  558.47  552.47
           MSFT    37.22   36.60
           AAPL    79.10   77.20
           AMZN   402.71  396.22

In [ ]:
tech.sort_index().loc['02-01-2014': '05-01-2014', 'close':'low']

close     high      low
date       name                            
02-01-2014 AAPL     79.02    79.58    78.86
           AMZN    397.97   399.36   394.02
           FB       54.71    55.22    54.19
           GOOGL   556.56   558.88   554.13
           MSFT     37.16    37.40    37.10
...                   ...      ...      ...
04-12-2018 AAPL    176.69   182.39   176.27
           AMZN   1668.40  1770.34  1665.00
           FB      137.93   143.39   137.16
           GOOGL  1062.47  1114.24  1060.78
           MSFT    108.52   112.64   108.21

[670 rows x 3 columns]

In [ ]:
tech.sort_index().loc[(slice('02-01-2014' , '05-01-2014'), ['GOOGL', 'MSFT']), 'close': 'low']

close     high      low
date       name                            
02-01-2014 GOOGL   556.56   558.88   554.13
           MSFT     37.16    37.40    37.10
02-01-2015 GOOGL   529.55   535.80   527.88
           MSFT     46.76    47.42    46.54
02-01-2018 GOOGL  1073.21  1075.98  1053.02
...                   ...      ...      ...
04-12-2015 MSFT     55.91    56.23    54.10
04-12-2017 GOOGL  1011.87  1031.34  1009.22
           MSFT     81.08    84.43    80.70
04-12-2018 GOOGL  1062.47  1114.24  1060.78
           MSFT    108.52   112.64   108.21

[268 rows x 3 columns]

In [ ]:
tech.sort_index().loc[(slice(None), ['GOOGL', 'MSFT']), 'close']

date        name 
01-02-2016  GOOGL     770.77
            MSFT       54.71
01-02-2017  GOOGL     815.24
            MSFT       63.58
01-02-2018  GOOGL    1181.59
                      ...   
31-12-2014  MSFT       46.45
31-12-2015  GOOGL     778.01
            MSFT       55.48
31-12-2018  GOOGL    1044.96
            MSFT      101.57
Name: close, Length: 2842, dtype: float64

### Lecture - 176 - Use : With pd.IndexSlice!

In [ ]:
i = pd.IndexSlice

In [ ]:
# jan 6 - jan 10
tech.sort_index().loc[i['01-02-2016': '05-02-2016', ['FB', 'GOOGL']], 'high':'low']

high     low
date       name                 
01-02-2016 FB     115.72  112.01
           GOOGL  777.59  761.38
01-02-2017 FB     133.49  130.68
           GOOGL  824.00  812.25
01-02-2018 FB     195.32  187.89
...                  ...     ...
05-02-2014 GOOGL  575.39  564.01
05-02-2015 FB      75.98   75.21
           GOOGL  530.69  525.64
05-02-2016 FB     109.58  103.18
           GOOGL  726.72  700.14

[368 rows x 2 columns]

In [ ]:
tech.sort_index().loc[i[:, ['FB', 'GOOGL']], 'high':'low']

high      low
date       name                   
01-02-2016 FB      115.72   112.01
           GOOGL   777.59   761.38
01-02-2017 FB      133.49   130.68
           GOOGL   824.00   812.25
01-02-2018 FB      195.32   187.89
...                   ...      ...
31-12-2014 GOOGL   538.40   530.20
31-12-2015 FB      106.17   104.62
           GOOGL   788.33   777.32
31-12-2018 FB      134.64   129.95
           GOOGL  1062.99  1033.04

[2842 rows x 2 columns]

### Lecture - 177 - Cross Sections with xs()

In [ ]:
tech.sort_index().loc[(slice(None), 'FB'), :]

month  year  day    open   close    high     low    volume  \
date       name                                                               
01-02-2016 FB        2  2016    1  112.27  115.09  115.72  112.01  46132726   
01-02-2017 FB        2  2017    1  132.25  133.23  133.49  130.68  50139775   
01-02-2018 FB        2  2018    1  188.22  193.09  195.32  187.89  54211293   
01-02-2019 FB        2  2019    1  165.84  165.71  169.10  165.66  30806513   
01-03-2016 FB        3  2016    1  107.83  109.82  109.82  107.60  26817309   
...                ...   ...  ...     ...     ...     ...     ...       ...   
31-10-2017 FB       10  2017   31  180.57  180.06  180.80  178.94  20174231   
31-10-2018 FB       10  2018   31  155.00  151.79  156.40  148.96  60101251   
31-12-2014 FB       12  2014   31   79.54   78.02   79.80   77.86  20040439   
31-12-2015 FB       12  2015   31  106.00  104.66  106.17  104.62  18391064   
31-12-2018 FB       12  2018   31  134.45  131.09  134.64  129.95  24625308   

                volume_type  
date       name              
01-02-2016 FB        medium  
01-02-2017 FB        medium  
01-02-2018 FB        medium  
01-02-2019 FB        medium  
01-03-2016 FB        medium  
...                     ...  
31-10-2017 FB        medium  
31-10-2018 FB          high  
31-12-2014 FB        medium  
31-12-2015 FB        medium  
31-12-2018 FB        medium  

[1421 rows x 9 columns]

In [ ]:
tech.xs('FB', level=1,drop_level= False)

month  year  day    open   close    high     low    volume  \
date       name                                                               
02-01-2014 FB        1  2014    2   54.86   54.71   55.22   54.19  43257622   
03-01-2014 FB        1  2014    3   55.00   54.56   55.65   54.53  38287706   
06-01-2014 FB        1  2014    6   54.39   57.20   57.26   54.05  68974359   
07-01-2014 FB        1  2014    7   57.67   57.92   58.55   57.22  77329009   
08-01-2014 FB        1  2014    8   57.59   58.23   58.41   57.23  56800776   
...                ...   ...  ...     ...     ...     ...     ...       ...   
19-08-2019 FB        8  2019   19  186.01  186.17  187.50  184.85   9699661   
20-08-2019 FB        8  2019   20  185.45  183.81  186.00  182.39  10087592   
21-08-2019 FB        8  2019   21  185.00  183.55  185.90  183.14   8409548   
22-08-2019 FB        8  2019   22  183.43  182.04  184.11  179.91  10829509   
23-08-2019 FB        8  2019   23  180.84  177.75  183.13  176.66  17331221   

                volume_type  
date       name              
02-01-2014 FB        medium  
03-01-2014 FB        medium  
06-01-2014 FB          high  
07-01-2014 FB          high  
08-01-2014 FB          high  
...                     ...  
19-08-2019 FB           low  
20-08-2019 FB           low  
21-08-2019 FB           low  
22-08-2019 FB           low  
23-08-2019 FB        medium  

[1421 rows x 9 columns]

In [ ]:
tech.xs(('02-01-2014': '04-01-2014', 'FB'), level = (0,1))

SyntaxError: invalid syntax (Temp/ipykernel_2460/481271172.py, line 1)

In [ ]:
tech.sort_index().loc[(slice(None), 'FB'), :]

In [ ]:
tech.sort_index().loc[(slice('02-01-2014', '04-01-2014'), 'FB'), :]

month  year  day    open   close    high     low    volume  \
date       name                                                               
02-01-2014 FB        1  2014    2   54.86   54.71   55.22   54.19  43257622   
02-01-2015 FB        1  2015    2   78.58   78.45   78.93   77.70  18177475   
02-01-2018 FB        1  2018    2  177.68  181.42  181.58  177.55  18151903   
02-01-2019 FB        1  2019    2  128.99  135.68  137.51  128.56  28146193   
02-02-2015 FB        2  2015    2   76.11   74.99   76.14   73.75  41955258   
...                ...   ...  ...     ...     ...     ...     ...       ...   
03-11-2016 FB       11  2016    3  122.00  120.00  123.28  119.50  63766259   
03-11-2017 FB       11  2017    3  179.29  178.92  179.86  176.71  17822131   
03-12-2014 FB       12  2014    3   75.39   74.88   75.73   74.40  16689857   
03-12-2015 FB       12  2015    3  106.27  104.38  106.85  103.35  23236485   
03-12-2018 FB       12  2018    3  143.00  141.09  143.68  140.76  24819226   

                volume_type  
date       name              
02-01-2014 FB        medium  
02-01-2015 FB        medium  
02-01-2018 FB        medium  
02-01-2019 FB        medium  
02-02-2015 FB        medium  
...                     ...  
03-11-2016 FB          high  
03-11-2017 FB        medium  
03-12-2014 FB        medium  
03-12-2015 FB        medium  
03-12-2018 FB        medium  

[92 rows x 9 columns]

### Lecture 179 and 180 - problem and solution

### Lecture 181 - The anatomy of Multi-Index

In [ ]:
tech.index

MultiIndex([('02-01-2014',    'FB'),
            ('02-01-2014',  'AAPL'),
            ('02-01-2014', 'GOOGL'),
            ('02-01-2014',  'MSFT'),
            ('02-01-2014',  'AMZN'),
            ('03-01-2014',    'FB'),
            ('03-01-2014', 'GOOGL'),
            ('03-01-2014',  'MSFT'),
            ('03-01-2014',  'AAPL'),
            ('03-01-2014',  'AMZN'),
            ...
            ('22-08-2019',  'MSFT'),
            ('22-08-2019',    'FB'),
            ('22-08-2019',  'AMZN'),
            ('22-08-2019',  'AAPL'),
            ('22-08-2019', 'GOOGL'),
            ('23-08-2019',  'MSFT'),
            ('23-08-2019',  'AAPL'),
            ('23-08-2019', 'GOOGL'),
            ('23-08-2019',  'AMZN'),
            ('23-08-2019',    'FB')],
           names=['date', 'name'], length=7105)

In [ ]:
# Names
tech.index.names

FrozenList(['date', 'name'])

In [ ]:
# Levels
tech.index.nlevels

2

In [ ]:
len(tech.index.levels)

2

In [ ]:
tech.index.levels[0]

Index(['01-02-2016', '01-02-2017', '01-02-2018', '01-02-2019', '01-03-2016',
       '01-03-2017', '01-03-2018', '01-03-2019', '01-04-2014', '01-04-2015',
       ...
       '31-08-2016', '31-08-2017', '31-08-2018', '31-10-2014', '31-10-2016',
       '31-10-2017', '31-10-2018', '31-12-2014', '31-12-2015', '31-12-2018'],
      dtype='object', name='date', length=1421)

In [ ]:
tech.index.levels[1]

Index(['AAPL', 'AMZN', 'FB', 'GOOGL', 'MSFT'], dtype='object', name='name')

In [ ]:
# levshape
tech.index.levshape

(1421, 5)

In [ ]:
# values
tech.index.values

array([('02-01-2014', 'FB'), ('02-01-2014', 'AAPL'),
       ('02-01-2014', 'GOOGL'), ..., ('23-08-2019', 'GOOGL'),
       ('23-08-2019', 'AMZN'), ('23-08-2019', 'FB')], dtype=object)

### Lecture 181 - Adding Another level

In [ ]:
tech.set_index('volume_type', append = True, inplace = True)

In [ ]:
tech.index.nlevels

3

In [ ]:
tech.index.levels[2]

Index(['high', 'low', 'medium'], dtype='object', name='volume_type')

In [ ]:
tech.index.values

array([('02-01-2014', 'FB', 'medium'), ('02-01-2014', 'AAPL', 'low'),
       ('02-01-2014', 'GOOGL', 'medium'), ...,
       ('23-08-2019', 'GOOGL', 'medium'),
       ('23-08-2019', 'AMZN', 'medium'), ('23-08-2019', 'FB', 'medium')],
      dtype=object)

In [ ]:
# Jan 2019, high-Volume trading days

In [ ]:
tech.sort_index().loc[(slice('01-01-2019', '31-01-2019'), slice(None), 'high'), :]

month  year  day     open    close     high  \
date       name  volume_type                                                
01-02-2016 GOOGL high             2  2016    1   771.26   770.77   777.59   
01-02-2017 AAPL  high             2  2017    1   127.03   128.75   130.49   
01-02-2018 AMZN  high             2  2018    1  1445.00  1390.00  1459.88   
01-02-2019 AMZN  high             2  2019    1  1638.88  1626.23  1673.06   
01-04-2014 FB    high             4  2014    1    60.49    62.62    62.66   
...                             ...   ...  ...      ...      ...      ...   
31-01-2014 FB    high             1  2014   31    60.25    62.57    63.37   
           GOOGL high             1  2014   31   585.67   590.49   593.27   
           MSFT  high             1  2014   31    36.95    37.84    37.89   
31-01-2019 AMZN  high             1  2019   31  1692.85  1718.73  1736.41   
           FB    high             1  2019   31   165.60   166.69   171.68   

                                  low     volume  
date       name  volume_type                      
01-02-2016 GOOGL high          761.38    5576089  
01-02-2017 AAPL  high          127.01  111985040  
01-02-2018 AMZN  high         1385.14    9113808  
01-02-2019 AMZN  high         1622.01   11506213  
01-04-2014 FB    high           60.24   59291210  
...                               ...        ...  
31-01-2014 FB    high           60.17   87930298  
           GOOGL high          575.57    5569794  
           MSFT  high           36.56   93190755  
31-01-2019 AMZN  high         1679.08   10910338  
           FB    high          165.00   77233602  

[372 rows x 8 columns]

In [ ]:
# All high-Volume trading days

In [ ]:
tech.xs(level=2, key='high', drop_level=False)

month  year  day     open    close     high  \
date       name  volume_type                                                
06-01-2014 FB    high             1  2014    6    54.39    57.20    57.26   
07-01-2014 FB    high             1  2014    7    57.67    57.92    58.55   
08-01-2014 FB    high             1  2014    8    57.59    58.23    58.41   
09-01-2014 FB    high             1  2014    9    58.66    57.22    58.96   
13-01-2014 FB    high             1  2014   13    57.89    55.91    58.25   
...                             ...   ...  ...      ...      ...      ...   
30-04-2019 GOOGL high             4  2019   30  1190.63  1198.96  1200.98   
03-06-2019 FB    high             6  2019    3   175.00   164.15   175.05   
           AMZN  high             6  2019    3  1760.01  1692.69  1766.29   
           GOOGL high             6  2019    3  1066.93  1038.74  1067.00   
26-07-2019 GOOGL high             7  2019   26  1228.00  1245.22  1268.39   

                                  low    volume  
date       name  volume_type                     
06-01-2014 FB    high           54.05  68974359  
07-01-2014 FB    high           57.22  77329009  
08-01-2014 FB    high           57.23  56800776  
09-01-2014 FB    high           56.65  92349222  
13-01-2014 FB    high           55.38  63106519  
...                               ...       ...  
30-04-2019 GOOGL high         1183.00   6658855  
03-06-2019 FB    high          161.01  56059609  
           AMZN  high         1672.00   9098708  
           GOOGL high         1027.03   4844480  
26-07-2019 GOOGL high         1228.00   6060795  

[375 rows x 8 columns]

In [ ]:
tech.xs(level = (2,1), key=('high', 'FB'), drop_level=False)

month  year  day    open   close    high     low  \
date       name volume_type                                                     
06-01-2014 FB   high             1  2014    6   54.39   57.20   57.26   54.05   
07-01-2014 FB   high             1  2014    7   57.67   57.92   58.55   57.22   
08-01-2014 FB   high             1  2014    8   57.59   58.23   58.41   57.23   
09-01-2014 FB   high             1  2014    9   58.66   57.22   58.96   56.65   
13-01-2014 FB   high             1  2014   13   57.89   55.91   58.25   55.38   
...                            ...   ...  ...     ...     ...     ...     ...   
31-10-2018 FB   high            10  2018   31  155.00  151.79  156.40  148.96   
19-12-2018 FB   high            12  2018   19  141.21  133.24  144.91  132.50   
21-12-2018 FB   high            12  2018   21  133.39  124.95  134.90  123.42   
31-01-2019 FB   high             1  2019   31  165.60  166.69  171.68  165.00   
03-06-2019 FB   high             6  2019    3  175.00  164.15  175.05  161.01   

                               volume  
date       name volume_type            
06-01-2014 FB   high         68974359  
07-01-2014 FB   high         77329009  
08-01-2014 FB   high         56800776  
09-01-2014 FB   high         92349222  
13-01-2014 FB   high         63106519  
...                               ...  
31-10-2018 FB   high         60101251  
19-12-2018 FB   high         57404894  
21-12-2018 FB   high         56901491  
31-01-2019 FB   high         77233602  
03-06-2019 FB   high         56059609  

[115 rows x 8 columns]

### Lecture 182 - Shuffling Level

In [ ]:
tech.swaplevel(2,1)

month  year  day     open    close     high  \
date       volume_type name                                                 
02-01-2014 medium      FB         1  2014    2    54.86    54.71    55.22   
           low         AAPL       1  2014    2    79.38    79.02    79.58   
           medium      GOOGL      1  2014    2   557.73   556.56   558.88   
                       MSFT       1  2014    2    37.35    37.16    37.40   
                       AMZN       1  2014    2   398.80   397.97   399.36   
...                             ...   ...  ...      ...      ...      ...   
23-08-2019 medium      MSFT       8  2019   23   137.19   133.39   138.35   
                       AAPL       8  2019   23   209.43   202.64   212.05   
                       GOOGL      8  2019   23  1185.17  1153.58  1195.67   
                       AMZN       8  2019   23  1793.03  1749.62  1804.90   
                       FB         8  2019   23   180.84   177.75   183.13   

                                  low    volume  
date       volume_type name                      
02-01-2014 medium      FB       54.19  43257622  
           low         AAPL     78.86   8398851  
           medium      GOOGL   554.13   1822719  
                       MSFT     37.10  30643745  
                       AMZN    394.02   2140246  
...                               ...       ...  
23-08-2019 medium      MSFT    132.80  38515386  
                       AAPL    201.00  46882843  
                       GOOGL  1150.00   1813141  
                       AMZN   1745.23   5277898  
                       FB      176.66  17331221  

[7105 rows x 8 columns]

In [ ]:
tech

month  year  day     open    close     high  \
date       name  volume_type                                                
02-01-2014 FB    medium           1  2014    2    54.86    54.71    55.22   
           AAPL  low              1  2014    2    79.38    79.02    79.58   
           GOOGL medium           1  2014    2   557.73   556.56   558.88   
           MSFT  medium           1  2014    2    37.35    37.16    37.40   
           AMZN  medium           1  2014    2   398.80   397.97   399.36   
...                             ...   ...  ...      ...      ...      ...   
23-08-2019 MSFT  medium           8  2019   23   137.19   133.39   138.35   
           AAPL  medium           8  2019   23   209.43   202.64   212.05   
           GOOGL medium           8  2019   23  1185.17  1153.58  1195.67   
           AMZN  medium           8  2019   23  1793.03  1749.62  1804.90   
           FB    medium           8  2019   23   180.84   177.75   183.13   

                                  low    volume  
date       name  volume_type                     
02-01-2014 FB    medium         54.19  43257622  
           AAPL  low            78.86   8398851  
           GOOGL medium        554.13   1822719  
           MSFT  medium         37.10  30643745  
           AMZN  medium        394.02   2140246  
...                               ...       ...  
23-08-2019 MSFT  medium        132.80  38515386  
           AAPL  medium        201.00  46882843  
           GOOGL medium       1150.00   1813141  
           AMZN  medium       1745.23   5277898  
           FB    medium        176.66  17331221  

[7105 rows x 8 columns]

In [ ]:
tech = tech.swaplevel('volume_type', 'name')

In [ ]:
tech.head()

month  year  day    open   close    high  \
date       volume_type name                                              
02-01-2014 medium      FB         1  2014    2   54.86   54.71   55.22   
           low         AAPL       1  2014    2   79.38   79.02   79.58   
           medium      GOOGL      1  2014    2  557.73  556.56  558.88   
                       MSFT       1  2014    2   37.35   37.16   37.40   
                       AMZN       1  2014    2  398.80  397.97  399.36   

                                 low    volume  
date       volume_type name                     
02-01-2014 medium      FB      54.19  43257622  
           low         AAPL    78.86   8398851  
           medium      GOOGL  554.13   1822719  
                       MSFT    37.10  30643745  
                       AMZN   394.02   2140246

In [ ]:
tech.reorder_levels([2,0,1])

month  year  day     open    close     high  \
name  date       volume_type                                                
FB    02-01-2014 medium           1  2014    2    54.86    54.71    55.22   
AAPL  02-01-2014 low              1  2014    2    79.38    79.02    79.58   
GOOGL 02-01-2014 medium           1  2014    2   557.73   556.56   558.88   
MSFT  02-01-2014 medium           1  2014    2    37.35    37.16    37.40   
AMZN  02-01-2014 medium           1  2014    2   398.80   397.97   399.36   
...                             ...   ...  ...      ...      ...      ...   
MSFT  23-08-2019 medium           8  2019   23   137.19   133.39   138.35   
AAPL  23-08-2019 medium           8  2019   23   209.43   202.64   212.05   
GOOGL 23-08-2019 medium           8  2019   23  1185.17  1153.58  1195.67   
AMZN  23-08-2019 medium           8  2019   23  1793.03  1749.62  1804.90   
FB    23-08-2019 medium           8  2019   23   180.84   177.75   183.13   

                                  low    volume  
name  date       volume_type                     
FB    02-01-2014 medium         54.19  43257622  
AAPL  02-01-2014 low            78.86   8398851  
GOOGL 02-01-2014 medium        554.13   1822719  
MSFT  02-01-2014 medium         37.10  30643745  
AMZN  02-01-2014 medium        394.02   2140246  
...                               ...       ...  
MSFT  23-08-2019 medium        132.80  38515386  
AAPL  23-08-2019 medium        201.00  46882843  
GOOGL 23-08-2019 medium       1150.00   1813141  
AMZN  23-08-2019 medium       1745.23   5277898  
FB    23-08-2019 medium        176.66  17331221  

[7105 rows x 8 columns]

In [ ]:
tech.index.reorder_levels([2,1,0])

MultiIndex([(   'FB', 'medium', '02-01-2014'),
            ( 'AAPL',    'low', '02-01-2014'),
            ('GOOGL', 'medium', '02-01-2014'),
            ( 'MSFT', 'medium', '02-01-2014'),
            ( 'AMZN', 'medium', '02-01-2014'),
            (   'FB', 'medium', '03-01-2014'),
            ('GOOGL', 'medium', '03-01-2014'),
            ( 'MSFT', 'medium', '03-01-2014'),
            ( 'AAPL',    'low', '03-01-2014'),
            ( 'AMZN', 'medium', '03-01-2014'),
            ...
            ( 'MSFT', 'medium', '22-08-2019'),
            (   'FB',    'low', '22-08-2019'),
            ( 'AMZN', 'medium', '22-08-2019'),
            ( 'AAPL', 'medium', '22-08-2019'),
            ('GOOGL',    'low', '22-08-2019'),
            ( 'MSFT', 'medium', '23-08-2019'),
            ( 'AAPL', 'medium', '23-08-2019'),
            ('GOOGL', 'medium', '23-08-2019'),
            ( 'AMZN', 'medium', '23-08-2019'),
            (   'FB', 'medium', '23-08-2019')],
           names=['name', 'volume_type', 'd

### Lecture 183 - Removing Multi-Index Levels

In [ ]:
tech.droplevel(1)

month  year  day     open    close     high      low  \
date       name                                                          
02-01-2014 FB         1  2014    2    54.86    54.71    55.22    54.19   
           AAPL       1  2014    2    79.38    79.02    79.58    78.86   
           GOOGL      1  2014    2   557.73   556.56   558.88   554.13   
           MSFT       1  2014    2    37.35    37.16    37.40    37.10   
           AMZN       1  2014    2   398.80   397.97   399.36   394.02   
...                 ...   ...  ...      ...      ...      ...      ...   
23-08-2019 MSFT       8  2019   23   137.19   133.39   138.35   132.80   
           AAPL       8  2019   23   209.43   202.64   212.05   201.00   
           GOOGL      8  2019   23  1185.17  1153.58  1195.67  1150.00   
           AMZN       8  2019   23  1793.03  1749.62  1804.90  1745.23   
           FB         8  2019   23   180.84   177.75   183.13   176.66   

                    volume  
date       name             
02-01-2014 FB     43257622  
           AAPL    8398851  
           GOOGL   1822719  
           MSFT   30643745  
           AMZN    2140246  
...                    ...  
23-08-2019 MSFT   38515386  
           AAPL   46882843  
           GOOGL   1813141  
           AMZN    5277898  
           FB     17331221  

[7105 rows x 8 columns]

In [ ]:
tech.reset_index(level=1, drop= False)

volume_type  month  year  day     open    close     high  \
date       name                                                             
02-01-2014 FB         medium      1  2014    2    54.86    54.71    55.22   
           AAPL          low      1  2014    2    79.38    79.02    79.58   
           GOOGL      medium      1  2014    2   557.73   556.56   558.88   
           MSFT       medium      1  2014    2    37.35    37.16    37.40   
           AMZN       medium      1  2014    2   398.80   397.97   399.36   
...                      ...    ...   ...  ...      ...      ...      ...   
23-08-2019 MSFT       medium      8  2019   23   137.19   133.39   138.35   
           AAPL       medium      8  2019   23   209.43   202.64   212.05   
           GOOGL      medium      8  2019   23  1185.17  1153.58  1195.67   
           AMZN       medium      8  2019   23  1793.03  1749.62  1804.90   
           FB         medium      8  2019   23   180.84   177.75   183.13   

                      low    volume  
date       name                      
02-01-2014 FB       54.19  43257622  
           AAPL     78.86   8398851  
           GOOGL   554.13   1822719  
           MSFT     37.10  30643745  
           AMZN    394.02   2140246  
...                   ...       ...  
23-08-2019 MSFT    132.80  38515386  
           AAPL    201.00  46882843  
           GOOGL  1150.00   1813141  
           AMZN   1745.23   5277898  
           FB      176.66  17331221  

[7105 rows x 9 columns]

In [ ]:
# How about removing several levels at once?

In [ ]:
tech.droplevel(['volume_type', 'name'])

month  year  day     open    close     high      low    volume
date                                                                      
02-01-2014      1  2014    2    54.86    54.71    55.22    54.19  43257622
02-01-2014      1  2014    2    79.38    79.02    79.58    78.86   8398851
02-01-2014      1  2014    2   557.73   556.56   558.88   554.13   1822719
02-01-2014      1  2014    2    37.35    37.16    37.40    37.10  30643745
02-01-2014      1  2014    2   398.80   397.97   399.36   394.02   2140246
...           ...   ...  ...      ...      ...      ...      ...       ...
23-08-2019      8  2019   23   137.19   133.39   138.35   132.80  38515386
23-08-2019      8  2019   23   209.43   202.64   212.05   201.00  46882843
23-08-2019      8  2019   23  1185.17  1153.58  1195.67  1150.00   1813141
23-08-2019      8  2019   23  1793.03  1749.62  1804.90  1745.23   5277898
23-08-2019      8  2019   23   180.84   177.75   183.13   176.66  17331221

[7105 rows x 8 columns]

In [ ]:
tech.reset_index(['volume_type', 'name'], drop = False)

volume_type   name  month  year  day     open    close     high  \
date                                                                         
02-01-2014      medium     FB      1  2014    2    54.86    54.71    55.22   
02-01-2014         low   AAPL      1  2014    2    79.38    79.02    79.58   
02-01-2014      medium  GOOGL      1  2014    2   557.73   556.56   558.88   
02-01-2014      medium   MSFT      1  2014    2    37.35    37.16    37.40   
02-01-2014      medium   AMZN      1  2014    2   398.80   397.97   399.36   
...                ...    ...    ...   ...  ...      ...      ...      ...   
23-08-2019      medium   MSFT      8  2019   23   137.19   133.39   138.35   
23-08-2019      medium   AAPL      8  2019   23   209.43   202.64   212.05   
23-08-2019      medium  GOOGL      8  2019   23  1185.17  1153.58  1195.67   
23-08-2019      medium   AMZN      8  2019   23  1793.03  1749.62  1804.90   
23-08-2019      medium     FB      8  2019   23   180.84   177.75   183.13   

                low    volume  
date                           
02-01-2014    54.19  43257622  
02-01-2014    78.86   8398851  
02-01-2014   554.13   1822719  
02-01-2014    37.10  30643745  
02-01-2014   394.02   2140246  
...             ...       ...  
23-08-2019   132.80  38515386  
23-08-2019   201.00  46882843  
23-08-2019  1150.00   1813141  
23-08-2019  1745.23   5277898  
23-08-2019   176.66  17331221  

[7105 rows x 10 columns]

In [ ]:
# what about total reset?

In [ ]:
tech.reset_index()

date volume_type   name  month  year  day     open    close  \
0     02-01-2014      medium     FB      1  2014    2    54.86    54.71   
1     02-01-2014         low   AAPL      1  2014    2    79.38    79.02   
2     02-01-2014      medium  GOOGL      1  2014    2   557.73   556.56   
3     02-01-2014      medium   MSFT      1  2014    2    37.35    37.16   
4     02-01-2014      medium   AMZN      1  2014    2   398.80   397.97   
...          ...         ...    ...    ...   ...  ...      ...      ...   
7100  23-08-2019      medium   MSFT      8  2019   23   137.19   133.39   
7101  23-08-2019      medium   AAPL      8  2019   23   209.43   202.64   
7102  23-08-2019      medium  GOOGL      8  2019   23  1185.17  1153.58   
7103  23-08-2019      medium   AMZN      8  2019   23  1793.03  1749.62   
7104  23-08-2019      medium     FB      8  2019   23   180.84   177.75   

         high      low    volume  
0       55.22    54.19  43257622  
1       79.58    78.86   8398851  
2      558.88   554.13   1822719  
3       37.40    37.10  30643745  
4      399.36   394.02   2140246  
...       ...      ...       ...  
7100   138.35   132.80  38515386  
7101   212.05   201.00  46882843  
7102  1195.67  1150.00   1813141  
7103  1804.90  1745.23   5277898  
7104   183.13   176.66  17331221  

[7105 rows x 11 columns]

### Lecture 184 - Multi-Index sort_index()

In [ ]:
tech.sort_index().loc[(slice('01-01-2014', '05-01-2014'), slice(None), slice('AAPL', 'FB')), 'open':'low']

open    close     high      low
date       volume_type name                                    
01-02-2016 medium      AAPL    96.47    96.43    96.71    95.40
                       AMZN   578.15   574.81   581.80   570.31
                       FB     112.27   115.09   115.72   112.01
01-02-2017 high        AAPL   127.03   128.75   130.49   127.01
           medium      AMZN   829.21   832.35   833.78   824.94
...                              ...      ...      ...      ...
04-12-2017 medium      AMZN  1173.85  1133.95  1175.20  1128.00
                       FB     176.29   171.47   176.57   170.79
04-12-2018 high        AMZN  1756.00  1668.40  1770.34  1665.00
           medium      AAPL   180.95   176.69   182.39   176.27
                       FB     140.73   137.93   143.39   137.16

[531 rows x 4 columns]

In [ ]:
# Sorting different levels in different orders

In [ ]:
tech.sort_index(level = (0,2), ascending= [False, True])

month  year  day     open    close     high  \
date       volume_type name                                                 
31-12-2018 medium      AAPL      12  2018   31   158.53   157.74   159.36   
                       AMZN      12  2018   31  1510.80  1501.97  1520.76   
                       FB        12  2018   31   134.45   131.09   134.64   
                       GOOGL     12  2018   31  1057.83  1044.96  1062.99   
                       MSFT      12  2018   31   101.29   101.57   102.40   
...                             ...   ...  ...      ...      ...      ...   
01-02-2016 medium      AAPL       2  2016    1    96.47    96.43    96.71   
                       AMZN       2  2016    1   578.15   574.81   581.80   
                       FB         2  2016    1   112.27   115.09   115.72   
           high        GOOGL      2  2016    1   771.26   770.77   777.59   
           medium      MSFT       2  2016    1    54.88    54.71    55.09   

                                  low    volume  
date       volume_type name                      
31-12-2018 medium      AAPL    156.48  35003466  
                       AMZN   1487.00   6954507  
                       FB      129.95  24625308  
                       GOOGL  1033.04   1655504  
                       MSFT    100.44  33173765  
...                               ...       ...  
01-02-2016 medium      AAPL     95.40  40943541  
                       AMZN    570.31   6355123  
                       FB      112.01  46132726  
           high        GOOGL   761.38   5576089  
           medium      MSFT     54.50  44208542  

[7105 rows x 8 columns]

### Lecture 185 - More Multi-Index Method

In [ ]:
tidx = tech.index

In [ ]:
tidx

MultiIndex([('02-01-2014', 'medium',    'FB'),
            ('02-01-2014',    'low',  'AAPL'),
            ('02-01-2014', 'medium', 'GOOGL'),
            ('02-01-2014', 'medium',  'MSFT'),
            ('02-01-2014', 'medium',  'AMZN'),
            ('03-01-2014', 'medium',    'FB'),
            ('03-01-2014', 'medium', 'GOOGL'),
            ('03-01-2014', 'medium',  'MSFT'),
            ('03-01-2014',    'low',  'AAPL'),
            ('03-01-2014', 'medium',  'AMZN'),
            ...
            ('22-08-2019', 'medium',  'MSFT'),
            ('22-08-2019',    'low',    'FB'),
            ('22-08-2019', 'medium',  'AMZN'),
            ('22-08-2019', 'medium',  'AAPL'),
            ('22-08-2019',    'low', 'GOOGL'),
            ('23-08-2019', 'medium',  'MSFT'),
            ('23-08-2019', 'medium',  'AAPL'),
            ('23-08-2019', 'medium', 'GOOGL'),
            ('23-08-2019', 'medium',  'AMZN'),
            ('23-08-2019', 'medium',    'FB')],
           names=['date', 'volume_type', 'n

In [ ]:
tidx.is_lexsorted()

C:\Users\JITENDRA\AppData\Local\Temp/ipykernel_24468/3114160297.py:1: FutureWarning: MultiIndex.is_lexsorted is deprecated as a public function, users should use MultiIndex.is_monotonic_increasing instead.
  tidx.is_lexsorted()


False

In [ ]:
type(tidx)

pandas.core.indexes.multi.MultiIndex

In [ ]:
tidx.sortlevel(0, ascending=False, sort_remaining=True)

(MultiIndex([('31-12-2018', 'medium',  'MSFT'),
             ('31-12-2018', 'medium', 'GOOGL'),
             ('31-12-2018', 'medium',    'FB'),
             ('31-12-2018', 'medium',  'AMZN'),
             ('31-12-2018', 'medium',  'AAPL'),
             ('31-12-2015', 'medium',  'MSFT'),
             ('31-12-2015', 'medium', 'GOOGL'),
             ('31-12-2015', 'medium',    'FB'),
             ('31-12-2015', 'medium',  'AMZN'),
             ('31-12-2015', 'medium',  'AAPL'),
             ...
             ('01-02-2017', 'medium',  'MSFT'),
             ('01-02-2017', 'medium', 'GOOGL'),
             ('01-02-2017', 'medium',    'FB'),
             ('01-02-2017', 'medium',  'AMZN'),
             ('01-02-2017',   'high',  'AAPL'),
             ('01-02-2016', 'medium',  'MSFT'),
             ('01-02-2016', 'medium',    'FB'),
             ('01-02-2016', 'medium',  'AMZN'),
             ('01-02-2016', 'medium',  'AAPL'),
             ('01-02-2016',   'high', 'GOOGL')],
            names=['da

In [ ]:
tidx.sortlevel((0,1,2), ascending = [True, True, False])

(MultiIndex([('01-02-2016',   'high', 'GOOGL'),
             ('01-02-2016', 'medium',  'MSFT'),
             ('01-02-2016', 'medium',    'FB'),
             ('01-02-2016', 'medium',  'AMZN'),
             ('01-02-2016', 'medium',  'AAPL'),
             ('01-02-2017',   'high',  'AAPL'),
             ('01-02-2017', 'medium',  'MSFT'),
             ('01-02-2017', 'medium', 'GOOGL'),
             ('01-02-2017', 'medium',    'FB'),
             ('01-02-2017', 'medium',  'AMZN'),
             ...
             ('31-12-2015', 'medium',  'MSFT'),
             ('31-12-2015', 'medium', 'GOOGL'),
             ('31-12-2015', 'medium',    'FB'),
             ('31-12-2015', 'medium',  'AMZN'),
             ('31-12-2015', 'medium',  'AAPL'),
             ('31-12-2018', 'medium',  'MSFT'),
             ('31-12-2018', 'medium', 'GOOGL'),
             ('31-12-2018', 'medium',    'FB'),
             ('31-12-2018', 'medium',  'AMZN'),
             ('31-12-2018', 'medium',  'AAPL')],
            names=['da

In [ ]:
tidx.set_names(['Trading Date', 'Volume Category', 'Ticker'], inplace = True)

In [ ]:
tidx

MultiIndex([('02-01-2014', 'medium',    'FB'),
            ('02-01-2014',    'low',  'AAPL'),
            ('02-01-2014', 'medium', 'GOOGL'),
            ('02-01-2014', 'medium',  'MSFT'),
            ('02-01-2014', 'medium',  'AMZN'),
            ('03-01-2014', 'medium',    'FB'),
            ('03-01-2014', 'medium', 'GOOGL'),
            ('03-01-2014', 'medium',  'MSFT'),
            ('03-01-2014',    'low',  'AAPL'),
            ('03-01-2014', 'medium',  'AMZN'),
            ...
            ('22-08-2019', 'medium',  'MSFT'),
            ('22-08-2019',    'low',    'FB'),
            ('22-08-2019', 'medium',  'AMZN'),
            ('22-08-2019', 'medium',  'AAPL'),
            ('22-08-2019',    'low', 'GOOGL'),
            ('23-08-2019', 'medium',  'MSFT'),
            ('23-08-2019', 'medium',  'AAPL'),
            ('23-08-2019', 'medium', 'GOOGL'),
            ('23-08-2019', 'medium',  'AMZN'),
            ('23-08-2019', 'medium',    'FB')],
           names=['Trading Date', 'Volume C

In [ ]:
tech

month  year  day     open    close  \
Trading Date Volume Category Ticker                                       
02-01-2014   medium          FB          1  2014    2    54.86    54.71   
             low             AAPL        1  2014    2    79.38    79.02   
             medium          GOOGL       1  2014    2   557.73   556.56   
                             MSFT        1  2014    2    37.35    37.16   
                             AMZN        1  2014    2   398.80   397.97   
...                                    ...   ...  ...      ...      ...   
23-08-2019   medium          MSFT        8  2019   23   137.19   133.39   
                             AAPL        8  2019   23   209.43   202.64   
                             GOOGL       8  2019   23  1185.17  1153.58   
                             AMZN        8  2019   23  1793.03  1749.62   
                             FB          8  2019   23   180.84   177.75   

                                        high      low    volume  
Trading Date Volume Category Ticker                              
02-01-2014   medium          FB        55.22    54.19  43257622  
             low             AAPL      79.58    78.86   8398851  
             medium          GOOGL    558.88   554.13   1822719  
                             MSFT      37.40    37.10  30643745  
                             AMZN     399.36   394.02   2140246  
...                                      ...      ...       ...  
23-08-2019   medium          MSFT     138.35   132.80  38515386  
                             AAPL     212.05   201.00  46882843  
                             GOOGL   1195.67  1150.00   1813141  
                             AMZN    1804.90  1745.23   5277898  
                             FB       183.13   176.66  17331221  

[7105 rows x 8 columns]

In [131]:
tidx.to_flat_index()

NameError: name 'tidx' is not defined

In [130]:
tech.close

date        name 
02-01-2014  FB         54.71
            AAPL       79.02
            GOOGL     556.56
            MSFT       37.16
            AMZN      397.97
                      ...   
23-08-2019  MSFT      133.39
            AAPL      202.64
            GOOGL    1153.58
            AMZN     1749.62
            FB        177.75
Name: close, Length: 7105, dtype: float64

### Lecture 186 - Reshaping with stack()

In [ ]:
stacked = tech.stack()

In [ ]:
# Columns Convert into Rows
stacked.head(10)

Trading Date  Volume Category  Ticker  Previously a column Axis
02-01-2014    medium           FB      month                              1.00
                                       year                            2014.00
                                       day                                2.00
                                       open                              54.86
                                       close                             54.71
                                       high                              55.22
                                       low                               54.19
                                       volume                      43257622.00
              low              AAPL    month                              1.00
                                       year                            2014.00
dtype: float64

In [ ]:
type(stacked)

pandas.core.series.Series

In [ ]:
stacked.index.nlevels

3

In [ ]:
names = stacked.index.names

In [129]:
names

FrozenList(['date', 'name', None])

In [ ]:
stacked.index.set_names([*names[:-1], 'Previously a column Axis'], inplace = True)

In [ ]:
stacked

Trading Date  Volume Category  Ticker  Previously a column Axis
02-01-2014    medium           FB      month                              1.00
                                       year                            2014.00
                                       day                                2.00
                                       open                              54.86
                                       close                             54.71
                                                                      ...     
23-08-2019    medium           FB      open                             180.84
                                       close                            177.75
                                       high                             183.13
                                       low                              176.66
                                       volume                      17331221.00
Length: 56840, dtype: float64

### Lecture 187 - Flipside: Unstack()

In [ ]:
stacked.unstack()

Previously a column Axis             month    year   day     open    close  \
Trading Date Volume Category Ticker                                          
01-02-2016   high            GOOGL     2.0  2016.0   1.0   771.26   770.77   
             medium          AAPL      2.0  2016.0   1.0    96.47    96.43   
                             AMZN      2.0  2016.0   1.0   578.15   574.81   
                             FB        2.0  2016.0   1.0   112.27   115.09   
                             MSFT      2.0  2016.0   1.0    54.88    54.71   
...                                    ...     ...   ...      ...      ...   
31-12-2018   medium          AAPL     12.0  2018.0  31.0   158.53   157.74   
                             AMZN     12.0  2018.0  31.0  1510.80  1501.97   
                             FB       12.0  2018.0  31.0   134.45   131.09   
                             GOOGL    12.0  2018.0  31.0  1057.83  1044.96   
                             MSFT     12.0  2018.0  31.0   101.29   101.57   

Previously a column Axis                high      low      volume  
Trading Date Volume Category Ticker                                
01-02-2016   high            GOOGL    777.59   761.38   5576089.0  
             medium          AAPL      96.71    95.40  40943541.0  
                             AMZN     581.80   570.31   6355123.0  
                             FB       115.72   112.01  46132726.0  
                             MSFT      55.09    54.50  44208542.0  
...                                      ...      ...         ...  
31-12-2018   medium          AAPL     159.36   156.48  35003466.0  
                             AMZN    1520.76  1487.00   6954507.0  
                             FB       134.64   129.95  24625308.0  
                             GOOGL   1062.99  1033.04   1655504.0  
                             MSFT     102.40   100.44  33173765.0  

[7105 rows x 8 columns]

In [ ]:
stacked.unstack().unstack()

Previously a column Axis     month                            year          \
Ticker                        AAPL  AMZN    FB GOOGL  MSFT    AAPL    AMZN   
Trading Date Volume Category                                                 
01-02-2016   high              NaN   NaN   NaN   2.0   NaN     NaN     NaN   
             medium            2.0   2.0   2.0   NaN   2.0  2016.0  2016.0   
01-02-2017   high              2.0   NaN   NaN   NaN   NaN  2017.0     NaN   
             medium            NaN   2.0   2.0   2.0   2.0     NaN  2017.0   
01-02-2018   high              NaN   2.0   NaN   NaN   NaN     NaN  2018.0   
...                            ...   ...   ...   ...   ...     ...     ...   
31-10-2018   medium           10.0   NaN   NaN  10.0  10.0  2018.0     NaN   
31-12-2014   low               NaN  12.0   NaN   NaN   NaN     NaN  2014.0   
             medium           12.0   NaN  12.0  12.0  12.0  2014.0     NaN   
31-12-2015   medium           12.0  12.0  12.0  12.0  12.0  2015.0  2015.0   
31-12-2018   medium           12.0  12.0  12.0  12.0  12.0  2018.0  2018.0   

Previously a column Axis                              ...     low           \
Ticker                            FB   GOOGL    MSFT  ...    AAPL     AMZN   
Trading Date Volume Category                          ...                    
01-02-2016   high                NaN  2016.0     NaN  ...     NaN      NaN   
             medium           2016.0     NaN  2016.0  ...   95.40   570.31   
01-02-2017   high                NaN     NaN     NaN  ...  127.01      NaN   
             medium           2017.0  2017.0  2017.0  ...     NaN   824.94   
01-02-2018   high                NaN     NaN     NaN  ...     NaN  1385.14   
...                              ...     ...     ...  ...     ...      ...   
31-10-2018   medium              NaN  2018.0  2018.0  ...  216.62      NaN   
31-12-2014   low                 NaN     NaN     NaN  ...     NaN   310.01   
             medium           2014.0  2014.0  2014.0  ...  110.21      NaN   
31-12-2015   medium           2015.0  2015.0  2015.0  ...  104.82   675.89   
31-12-2018   medium           2018.0  2018.0  2018.0  ...  156.48  1487.00   

Previously a column Axis                                    volume             \
Ticker                            FB    GOOGL    MSFT         AAPL       AMZN   
Trading Date Volume Category                                                    
01-02-2016   high                NaN   761.38     NaN          NaN        NaN   
             medium           112.01      NaN   54.50   40943541.0  6355123.0   
01-02-2017   high                NaN      NaN     NaN  111985040.0        NaN   
             medium           130.68   812.25   63.47          NaN  3850181.0   
01-02-2018   high                NaN      NaN     NaN          NaN  9113808.0   
...                              ...      ...     ...          ...        ...   
31-10-2018   medium              NaN  1068.20  105.39   38358933.0        NaN   
31-12-2014   low                 NaN      NaN     NaN          NaN  2057766.0   
             medium            77.86   530.20   46.45   41403351.0        NaN   
31-12-2015   medium           104.62   777.32   55.42   40912316.0  3749560.0   
31-12-2018   medium           129.95  1033.04  100.44   35003466.0  6954507.0   

Previously a column Axis                                         
Ticker                                FB      GOOGL        MSFT  
Trading Date Volume Category                                     
01-02-2016   high                    NaN  5576089.0         NaN  
             medium           46132726.0        NaN  44208542.0  
01-02-2017   high                    NaN        NaN         NaN  
             medium           50139775.0  2251047.0  39671528.0  
01-02-2018   high                    NaN        NaN         NaN  
...                                  ...        ...         ...  
31-10-2018   medium                  NaN  3545821.0  51062383.0  
31-12-2014   low   

In [ ]:
stacked.unstack(level=1)

Volume Category                               high  low       medium
Trading Date Ticker Previously a column Axis                        
01-02-2016   AAPL   month                      NaN  NaN         2.00
                    year                       NaN  NaN      2016.00
                    day                        NaN  NaN         1.00
                    open                       NaN  NaN        96.47
                    close                      NaN  NaN        96.43
...                                            ...  ...          ...
31-12-2018   MSFT   open                       NaN  NaN       101.29
                    close                      NaN  NaN       101.57
                    high                       NaN  NaN       102.40
                    low                        NaN  NaN       100.44
                    volume                     NaN  NaN  33173765.00

[56840 rows x 3 columns]

In [ ]:
stacked.unstack(level='Ticker')

Ticker                                                        AAPL  \
Trading Date Volume Category Previously a column Axis                
01-02-2016   high            month                             NaN   
                             year                              NaN   
                             day                               NaN   
                             open                              NaN   
                             close                             NaN   
...                                                            ...   
31-12-2018   medium          open                           158.53   
                             close                          157.74   
                             high                           159.36   
                             low                            156.48   
                             volume                    35003466.00   

Ticker                                                       AMZN  \
Trading Date Volume Category Previously a column Axis               
01-02-2016   high            month                            NaN   
                             year                             NaN   
                             day                              NaN   
                             open                             NaN   
                             close                            NaN   
...                                                           ...   
31-12-2018   medium          open                         1510.80   
                             close                        1501.97   
                             high                         1520.76   
                             low                          1487.00   
                             volume                    6954507.00   

Ticker                                                          FB  \
Trading Date Volume Category Previously a column Axis                
01-02-2016   high            month                             NaN   
                             year                              NaN   
                             day                               NaN   
                             open                              NaN   
                             close                             NaN   
...                                                            ...   
31-12-2018   medium          open                           134.45   
                             close                          131.09   
                             high                           134.64   
                             low                            129.95   
                             volume                    24625308.00   

Ticker                                                      GOOGL         MSFT  
Trading Date Volume Category Previously a column Axis                           
01-02-2016   high            month                           2.00          NaN  
                             year                         2016.00          NaN  
                             day                             1.00          NaN  
                             open                          771.26          NaN  
                             close                         770.77          NaN  
...                                                           ...          ...  
31-12-2018   medium          open                         1057.83       101.29  
                             close                        1044.96       101.57  
                             high                         1062.99       102.40  
                             low                          1033.04       100.44  
                             volume                    1655504.00  33173765.00  

[16720 rows x 5 columns]

### Lecture 188 - Creating MultiLevel Columns Manually

In [ ]:
# Let's Build one Manually: 

# Criteria:
# - 2 - level multiindex column axis, containing:
# - trading Volume (high and low) and tickers (MSFT and AMZN)
# - 10 random closing price observations

In [ ]:
tech

month  year  day     open    close     high      low  \
date       name                                                          
02-01-2014 FB         1  2014    2    54.86    54.71    55.22    54.19   
           AAPL       1  2014    2    79.38    79.02    79.58    78.86   
           GOOGL      1  2014    2   557.73   556.56   558.88   554.13   
           MSFT       1  2014    2    37.35    37.16    37.40    37.10   
           AMZN       1  2014    2   398.80   397.97   399.36   394.02   
...                 ...   ...  ...      ...      ...      ...      ...   
23-08-2019 MSFT       8  2019   23   137.19   133.39   138.35   132.80   
           AAPL       8  2019   23   209.43   202.64   212.05   201.00   
           GOOGL      8  2019   23  1185.17  1153.58  1195.67  1150.00   
           AMZN       8  2019   23  1793.03  1749.62  1804.90  1745.23   
           FB         8  2019   23   180.84   177.75   183.13   176.66   

                    volume volume_type  
date       name                         
02-01-2014 FB     43257622      medium  
           AAPL    8398851         low  
           GOOGL   1822719      medium  
           MSFT   30643745      medium  
           AMZN    2140246      medium  
...                    ...         ...  
23-08-2019 MSFT   38515386      medium  
           AAPL   46882843      medium  
           GOOGL   1813141      medium  
           AMZN    5277898      medium  
           FB     17331221      medium  

[7105 rows x 9 columns]

In [ ]:
tech.reset_index(inplace = True)

In [ ]:
tech

date   name  month  year  day     open    close     high      low  \
0     02-01-2014     FB      1  2014    2    54.86    54.71    55.22    54.19   
1     02-01-2014   AAPL      1  2014    2    79.38    79.02    79.58    78.86   
2     02-01-2014  GOOGL      1  2014    2   557.73   556.56   558.88   554.13   
3     02-01-2014   MSFT      1  2014    2    37.35    37.16    37.40    37.10   
4     02-01-2014   AMZN      1  2014    2   398.80   397.97   399.36   394.02   
...          ...    ...    ...   ...  ...      ...      ...      ...      ...   
7100  23-08-2019   MSFT      8  2019   23   137.19   133.39   138.35   132.80   
7101  23-08-2019   AAPL      8  2019   23   209.43   202.64   212.05   201.00   
7102  23-08-2019  GOOGL      8  2019   23  1185.17  1153.58  1195.67  1150.00   
7103  23-08-2019   AMZN      8  2019   23  1793.03  1749.62  1804.90  1745.23   
7104  23-08-2019     FB      8  2019   23   180.84   177.75   183.13   176.66   

        volume volume_type  
0     43257622      medium  
1      8398851         low  
2      1822719      medium  
3     30643745      medium  
4      2140246      medium  
...        ...         ...  
7100  38515386      medium  
7101  46882843      medium  
7102   1813141      medium  
7103   5277898      medium  
7104  17331221      medium  

[7105 rows x 11 columns]